channeldb/messagestore.go

```diff
@@ -0,0 +1,164 @@
+func messageStoreKey(msg lnwire.Message, peerPubKey [33]byte) ([]byte, error) {
+	var shortChanID lnwire.ShortChannelID
+	switch msg := msg.(type) {
+	case *lnwire.AnnounceSignatures:
+		shortChanID = msg.ShortChannelID
+	case *lnwire.ChannelUpdate:
+		shortChanID = msg.ShortChannelID
+	default:
+		return nil, ErrUnsupportedMessage
+	}
+
+	var k [33 + 8 + 2]byte
+	copy(k[:33], peerPubKey[:])
+	byteOrder.PutUint64(k[33:41], shortChanID.ToUint64())
+	byteOrder.PutUint16(k[41:43], uint16(msg.MsgType()))
+
+	return k[:], nil
+}
+
+// AddMessage adds a message to the store for this peer.
+func (s *GossipMessageStore) AddMessage(msg lnwire.Message,
+	peerPubKey [33]byte) error {
+
+	// Construct the key for which we'll find this message with in the store.
+	msgKey, err := messageStoreKey(msg, peerPubKey)
+	if err != nil {
+		return err
+	}
+
+	// Serialize the message with its wire encoding.
+	var b bytes.Buffer
+	if _, err := lnwire.WriteMessage(&b, msg, 0); err != nil {
+		return err
+	}
+
+	return s.db.Update(func(tx *bbolt.Tx) error {
+		messageStore, err := tx.CreateBucketIfNotExists(messageStoreBucket)
+		if err != nil {
+			return err
+		}
+		return messageStore.Put(msgKey, b.Bytes())
+	})
+}
+
+// DeleteMessage deletes a message from the store for this peer.
+func (s *GossipMessageStore) DeleteMessage(msg lnwire.Message,
+	peerPubKey [33]byte) error {
+
+	// Construct the key for which we'll find this message with in the store.
+	msgKey, err := messageStoreKey(msg, peerPubKey)
+	if err != nil {
+		return err
+	}
+
+	return s.db.Update(func(tx *bbolt.Tx) error {
+		messageStore := tx.Bucket(messageStoreBucket)
+		if messageStore == nil {
+			return errors.New("message store does not contain any " +
+				"messages")
+		}
+		return messageStore.Delete(msgKey)
+	})
+}
+
+// Messages returns the total set of messages that exist within the store for
+// all peers.
+func (s *GossipMessageStore) Messages() (map[[33]byte][]lnwire.Message, error) {
+	msgs := make(map[[33]byte][]lnwire.Message)
+	err := s.db.View(func(tx *bbolt.Tx) error {
+		messageStore := tx.Bucket(messageStoreBucket)
+		if messageStore == nil {
+			return nil
+		}
+
+		return messageStore.ForEach(func(k, v []byte) error {
+			var pubKey [33]byte
+			copy(pubKey[:], k[:33])
+			msg, err := lnwire.ReadMessage(bytes.NewReader(v), 0)
+			if err != nil {
+				return err
+			}
+
+			msgs[pubKey] = append(msgs[pubKey], msg)
+			return nil
+		})
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return msgs, nil
+}
+
+// MessagesForPeer returns the set of messages that exists within the store for
+// the given peer.
+func (s *GossipMessageStore) MessagesForPeer(
+	peerPubKey [33]byte) ([]lnwire.Message, error) {
+
+	var msgs []lnwire.Message
+	err := s.db.View(func(tx *bbolt.Tx) error {
+		messageStore := tx.Bucket(messageStoreBucket)
+		if messageStore == nil {
+			return nil
+		}
+
+		return messageStore.ForEach(func(k, v []byte) error {
```

Comment:
since all messages for the same peer should be adjacent, we should be able to use `Seek` instead of iterating over all stored messages

Reply:
Fixed.